lib/dRep/dRepImpl.py

```diff
@@ -80,102 +77,321 @@ def run_dRep(self, ctx, params):
-            bins_dir = '/kb/module/test/data/binned_contig_files_8bins'
+        #
+        if _globals.debug and params.get('skip_dl'):
+            bins_dir_name_l = params['bins_dir_name_l']
+            bins_dir_l = [os.path.join(_globals.shared_folder, bins_dir_name) for bins_dir_name in bins_dir_name_l]
 
+            for upa, bins_dir in zip(params['genomes_refs'], bins_dir_l):
+                BinnedContigs(upa, get_bins_dir='local', bins_dir=bins_dir)
+
+
+        # load from KBase 
         else:
 
-            binnedContigs_mguObjData = self.mgu.binned_contigs_to_file({'input_ref': binnedContigs_upa, 'save_to_shock': 0})
-            bins_dir = binnedContigs_mguObjData['bin_file_directory']
+            for upa in params['genomes_refs']:
+                bc = BinnedContigs(upa, get_bins_dir='download')
+                
+                if bc.is_empty():
+                    # never mention this bc again
+                    params['genomes_refs'].remove(upa)
+                    BinnedContigs.created_instances.remove(bc)
+                
 
-            dprint('binnedContigs_mguObjData', binnedContigs_mguObjData)
-            
-        binnedContigs_wsObjData = self.ws.get_objects2({'objects':[{'ref': binnedContigs_upa}]})
-        assembly_upa = binnedContigs_wsObjData['data'][0]['data']['assembly_ref']
-        
-        dprint('binnedContigs_wsObjData', binnedContigs_wsObjData)
+            if len(params['genomes_refs']) == 0:
+                msg = 'Sorry, please input at least one non-empty BinnedContigs'
+                raise ValueError(msg)
 
-    #        binnedContigs_wsObjInfo = self.ws.get_object_info([{'ref': binnedContigs_upa}])[0]
-    #        dprint('binnedContigs_wsObjInfo', binnedContigs_wsObjInfo)
 
 
         #
         ##
-        ### Run dRep dereplicate
+        ###
         ####
         #####
 
+        for bc in BinnedContigs.created_instances:
+            bc.calc_stats()
+
+
 
-        dRep_workDir = os.path.join(self.shared_folder, 'dRep_workDir_' + self.suffix)
+        #
+        ##
+        ### pool
+        ####
+        #####
+        
+ 
+        binsPooled_dir = os.path.join(_globals.run_dir, 'binsPooled')
+        os.mkdir(binsPooled_dir)
 
-        dRep_cmd = f'dRep dereplicate {dRep_workDir} -g {bins_dir}/*.fasta' 
-        if 'mode' in params and params['mode'] == 'local':
-            dRep_cmd += ' --debug --checkM_method taxonomy_wf'
-        dprint(f'Running CMD: {dRep_cmd}')
+        for bc in BinnedContigs.created_instances:
+            bc.pool_into(binsPooled_dir)
         
+        dprint("os.listdir(binsPooled_dir)", run={**locals(), **globals()})
+
+
+        #
+        ##
+        ### 
+        #### params
+        #####
+       
+
+        # flatten param groups (if any)
+
+        paramGrp_key_l = ['filtering', 'genome_comparison', 'clustering', 'scoring', 'warnings']
+
+        for paramGrp_key in paramGrp_key_l:
+            if paramGrp_key in params:
```

Comment:
I would shorten these three lines to `for paramIndiv_key in params.get(paramGrp_key, {}): ...`